### PR TITLE
[#590][part-1] ManagedBuffer instead ByteBuf to hold ShuffleData 

### PR DESCRIPTION
**Pull request by xumanbu**
### What changes were proposed in this pull request?

part-1: 
1. add a ManagedBuffer instead ByteBuf to hold ShuffleData.
2. ShuffleResultData & GetLocalShuffleDataResponse use ManagedBuffer instead of ByteBuf
3. MessageEncoder write support sendfile

### Why are the changes needed?
Fix: #590 

### Does this PR introduce _any_ user-facing change?

(Please list the user-facing changes introduced by your change, including
No.

### How was this patch tested?
Integration Testing

### CODE REVIEW

**Comment by jerqi** [comment]
@leixm Could you help me review this pr?

**Comment by leixm** [comment]
> @leixm Could you help me review this pr?

Sure.

File: common/src/main/java/org/apache/uniffle/common/netty/protocol/GetLocalShuffleDataResponse.java [hunk 1]
```diff
@@ -18,53 +18,64 @@
...
+    return super.encodedLength() + Integer.BYTES + buffer.size();
   }
 
   @Override
   public void encode(ByteBuf buf) {
     super.encode(buf);
-    ByteBufUtils.copyByteBuf(data, buf);
-    data.release();
+    if (buffer instanceof FileSegmentManagedBuffer) {
+      buf.writeInt(buffer.size());
```
Comment (jerqi) [inline common/src/main/java/org/apache/uniffle/common/netty/protocol/GetLocalShuffleDataResponse.java:47]: Why do we only write the size of buffer?

File: common/src/main/java/org/apache/uniffle/common/netty/MessageEncoder.java [hunk 2]
```diff
@@ -59,5 +59,9 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
       byteBuf.release();
     }
     ctx.writeAndFlush(byteBuf);
+    // do transferTo send data after encode buffer send.
+    if (message instanceof Transferable) {
+      ((Transferable) message).transferTo(ctx.channel());
```
Comment (xumanbu) [inline common/src/main/java/org/apache/uniffle/common/netty/MessageEncoder.java:64]: @jerqi the line is write the FileSegmentManageBuffer ShuffleData.

File: common/src/main/java/org/apache/uniffle/common/netty/protocol/GetLocalShuffleDataResponse.java (hunk 1 above)
Comment (jerqi) [reply common/src/main/java/org/apache/uniffle/common/netty/protocol/GetLocalShuffleDataResponse.java:47]: How do the other systems like Spark, Celeborn implement this?

File: common/src/main/java/org/apache/uniffle/common/netty/MessageEncoder.java (hunk 2 above)
Comment (jerqi) [reply common/src/main/java/org/apache/uniffle/common/netty/MessageEncoder.java:64]: OK.

File: common/src/main/java/org/apache/uniffle/common/ShuffleDataResult.java [hunk 3]
```diff
@@ -54,28 +61,32 @@ public ShuffleDataResult(byte[] data, List<BufferSegment> bufferSegments) {
...
+    if (buffer == null) {
       return null;
     }
-    if (data.hasArray()) {
-      return data.array();
+    if (buffer.nioByteBuffer().hasArray()) {
+      return buffer.nioByteBuffer().array();
     }
-    return ByteBufUtils.readBytes(data);
+    return ByteBufUtils.readBytes(Unpooled.wrappedBuffer(buffer.nioByteBuffer()));
```
Comment (leixm) [inline common/src/main/java/org/apache/uniffle/common/ShuffleDataResult.java:70]: Maybe we can `return ByteBufUtils.readBytes(buffer.byteBuf());`

File: common/src/main/java/org/apache/uniffle/common/ShuffleDataResult.java [hunk 4]
```diff
@@ -54,28 +61,32 @@ public ShuffleDataResult(byte[] data, List<BufferSegment> bufferSegments) {
...
   }
 
   public ByteBuf getDataBuf() {
-    return data;
+    return Unpooled.wrappedBuffer(buffer.nioByteBuffer());
```
Comment (leixm) [inline common/src/main/java/org/apache/uniffle/common/ShuffleDataResult.java:74]: ditto.

**Comment by leixm** [comment]
Maybe we can also make `SendShuffleDataRequest` hold `ManagedBuffer` instead of `ByteBuf`.

File: common/src/main/java/org/apache/uniffle/common/netty/protocol/GetLocalShuffleDataResponse.java (hunk 1 above)
Comment (xumanbu) [reply common/src/main/java/org/apache/uniffle/common/netty/protocol/GetLocalShuffleDataResponse.java:47]: In order to keep the uniffle interface design as much as possible, reference Celeborn add ManagedBuffer to hold different type ShuffleData(bytebuf or FileRegion).

File: common/src/main/java/org/apache/uniffle/common/ShuffleDataResult.java (hunk 3 above)
Comment (xumanbu) [reply common/src/main/java/org/apache/uniffle/common/ShuffleDataResult.java:70]: ok,I' ll fix it

File: common/src/main/java/org/apache/uniffle/common/netty/protocol/GetLocalShuffleDataResponse.java (hunk 1 above)
Comment (jerqi) [reply common/src/main/java/org/apache/uniffle/common/netty/protocol/GetLocalShuffleDataResponse.java:47]: > In order to keep the uniffle interface design as much as possible, reference Celeborn&Spark add ManagedBuffer to hold different type ShuffleData(bytebuf or FileRegion).

Ok for me.

**Comment by jerqi** [comment]
> Maybe we can also make `SendShuffleDataRequest` hold `ManagedBuffer` instead of `ByteBuf`.

What's the benefit?

File: common/src/main/java/org/apache/uniffle/common/ShuffleDataResult.java (hunk 3 above)
Comment (jerqi) [reply common/src/main/java/org/apache/uniffle/common/ShuffleDataResult.java:70]: @xumanbu remind to fix this.

Comment (xumanbu) [reply common/src/main/java/org/apache/uniffle/common/ShuffleDataResult.java]: ok

**Review by leixm — APPROVED** [review]
+1 LGTM.

**Review by jerqi — APPROVED** [review]
LGTM, merged to master.

**Pull request merged**